### PR TITLE
Unit tests for redis register/unregister.

### PR DESCRIPTION
**Pull request by Jason-RH**
Adding unit tests to the redis_connector and testing out the miniredis test server (fake redis server for testing)

### CODE REVIEW

File: internal/controller/redis_connector_test.go [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+		expected []string
+	}{
+		{account: "01", nodeID: "node-a", expected: []string{"01:node-b:" + testHost}},
+		{account: "01", nodeID: "node-not-found", expected: []string{"01:node-b:" + testHost}},
+		{account: "01", nodeID: "node-b", expected: []string{}},
+	}
+
+	for _, tc := range tests {
+		UnregisterWithRedis(c, tc.account, tc.nodeID)
+		if got := c.SMembers("connections"); !reflect.DeepEqual(got.Val(), tc.expected) {
```
Comment (dehort) [inline internal/controller/redis_connector_test.go:72]: Should this verify that the "01:node-a" and "01:node-b" keys no longer exist?

File: internal/controller/redis_connector_test.go [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+	}
+}
+
+func TestGetRedisConnectionsByAccount(t *testing.T) {
+	s, _ := miniredis.Run()
+	defer s.Close()
+
+	c := newTestRedisClient(s.Addr())
+
+	c.SAdd("01", "node-a:localhost", "node-b:localhost")
```
Comment (dehort) [inline internal/controller/redis_connector_test.go:191]: I would use the Register method to setup this test data.

File: internal/controller/redis_connector_test.go [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+	}
+}
+
+func TestGetRedisConnectionsByHost(t *testing.T) {
+	s, _ := miniredis.Run()
+	defer s.Close()
+
+	c := newTestRedisClient(s.Addr())
+
+	c.SAdd("localhost", "01:node-a", "02:node-b", "01:node-b")
```
Comment (dehort) [inline internal/controller/redis_connector_test.go:217]: Like above, I would use the Register method to setup this test data.

File: internal/controller/redis_connector_test.go [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+	}
+}
+
+func TestGetAllRedisConnections(t *testing.T) {
+	s, _ := miniredis.Run()
+	defer s.Close()
+
+	c := newTestRedisClient(s.Addr())
+
+	c.SAdd("connections", "01:node-a:localhost", "02:node-b:localhost", "01:node-b:localhost")
```
Comment (dehort) [inline internal/controller/redis_connector_test.go:243]: Like above, I would use the Register method to setup this test data.

File: internal/controller/redis_connector_test.go [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+	tests := []struct {
+		account string
+		nodeID  string
+	}{
+		{account: "01", nodeID: "node-a"},
+		{account: "01", nodeID: "node-b"},
+	}
+
+	for _, tc := range tests {
+		UnregisterWithRedis(c, tc.account, tc.nodeID)
```
Comment (dehort) [inline internal/controller/redis_connector_test.go:152]: I think we probably need to add explicit checks to make sure the that only the target node is removed.

File: internal/controller/redis_connector_test.go [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+	assert.Equal(t, c.SMembers(testHost).Val(), []string{})
+}
+
+func TestGetRedisConnection(t *testing.T) {
+	s, _ := miniredis.Run()
+	defer s.Close()
+
+	c := newTestRedisClient(s.Addr())
+
+	c.Set("01:node-a", "localhost", 0)
```
[the remark this answers is not in the document]
Comment (Jason-RH) [reply internal/controller/redis_connector_test.go:166]: Ah yeah that's a good point. I'll make that change throughout the tests.

File: internal/controller/redis_connector_test.go [hunk 7]
```diff
@@ -0,0 +1,280 @@
...
+
+	for _, tc := range tests {
+		got := RegisterWithRedis(c, tc.account, tc.nodeID, tc.hostname)
+		if got != tc.err {
+			t.Fatalf("expected: %v, got: %v", tc.err, got)
+		}
+		assert.Equal(t, c.Get(tc.account+":"+tc.nodeID).Val(), tc.expectedHost)
+		assert.Equal(t, c.SMembers(tc.account).Val(), tc.accIndex)
+		assert.Equal(t, c.SMembers("connections").Val(), tc.connIndex)
+		assert.Equal(t, c.SMembers(tc.hostname).Val(), tc.podIndex)
```
Comment (dehort) [inline internal/controller/redis_connector_test.go:145]: This tests that a new pod entry is not made for the duplicate connection.  This is certainly needed, but I think we should also make sure that the original pod entry has not been damaged.  What do you think?

File: internal/controller/redis_connector_test.go [hunk 8]
```diff
@@ -0,0 +1,280 @@
...
+	for _, tc := range tests {
+		UnregisterWithRedis(c, tc.account, tc.nodeID, testHost)
+		assert.Equal(t, c.Get(tc.account+":"+tc.nodeID).Err(), redis.Nil)
+		assert.Equal(t, c.SMembers(testHost).Val(), tc.remaining)
+	}
+
+	assert.Equal(t, c.SMembers("01").Val(), []string{})
+	assert.Equal(t, c.SMembers("connections").Val(), []string{})
+	assert.Equal(t, c.SMembers(testHost).Val(), []string{})
+}
```
Comment (dehort) [inline internal/controller/redis_connector_test.go:179]: This tests makes sure that all the registered connections have been cleaned up.  What about registering a connection that does not get removed?  The test could make sure that the connection that does not get unregistered is not damaged?

File: internal/controller/redis_connector.go [hunk 9]
```diff
@@ -4,11 +4,10 @@ import (
 	"strings"
 
 	"github.com/RedHatInsights/platform-receptor-controller/internal/platform/logger"
-	"github.com/RedHatInsights/platform-receptor-controller/internal/platform/utils"
 	"github.com/go-redis/redis"
 )
 
-var hostname string = utils.GetHostname()
+//var hostname string = utils.GetHostname()
```
[the remark this answers is not in the document]
Comment (Jason-RH) [reply internal/controller/redis_connector.go:10]: Yep :)

File: internal/controller/redis_connector_test.go (hunk 8 above)
Comment (Jason-RH) [reply internal/controller/redis_connector_test.go:179]: Sure I can add that in.

File: internal/controller/redis_connector_test.go (hunk 7 above)
Comment (Jason-RH) [reply internal/controller/redis_connector_test.go:145]: By the original pod entry do you mean the state of the pod index? If so I can add an additional check to make sure that the index for the other pod wasn't modified as part of this registration.

Comment (dehort) [reply internal/controller/redis_connector_test.go:145]: Yep

**Pull request merged**